scheduler/feasible.go

```diff
@@ -480,3 +450,113 @@ func checkRegexpConstraint(ctx Context, lVal, rVal interface{}) bool {
 	// Look for a match
 	return re.MatchString(lStr)
 }
+
+// FeasibilityWrapper is a FeasibleIterator which wraps both job and task group
+// FeasibilityCheckers in which feasibility checking can be skipped if the
+// computed node class has previously been marked as eligible or ineligible.
+type FeasibilityWrapper struct {
+	ctx         Context
+	source      FeasibleIterator
+	jobCheckers []FeasibilityChecker
+	tgCheckers  []FeasibilityChecker
+	tg          string
+}
+
+// NewFeasibilityWrapper returns a FeasibleIterator based on the passed source
+// and FeasibilityCheckers.
+func NewFeasibilityWrapper(ctx Context, source FeasibleIterator,
+	jobCheckers, tgCheckers []FeasibilityChecker) *FeasibilityWrapper {
+	return &FeasibilityWrapper{
+		ctx:         ctx,
+		source:      source,
+		jobCheckers: jobCheckers,
+		tgCheckers:  tgCheckers,
+	}
+}
+
+func (w *FeasibilityWrapper) SetTaskGroup(tg string) {
+	w.tg = tg
+}
+
+func (w *FeasibilityWrapper) Reset() {
+	w.source.Reset()
+}
+
+// Next returns an eligible node, only running the FeasibilityCheckers as needed
+// based on the sources computed node class.
+func (w *FeasibilityWrapper) Next() *structs.Node {
+	evalElig := w.ctx.Eligibility()
+	metrics := w.ctx.Metrics()
+
+OUTER:
+	for {
+		// Get the next option from the source
+		option := w.source.Next()
+		if option == nil {
+			return nil
+		}
+
+		// Check if the job has been marked as eligible or ineligible.
+		jobEscaped := false
+		switch evalElig.JobStatus(option.ComputedClass) {
+		case EvalComputedClassIneligible:
+			// Fast path the ineligible case
+			metrics.FilterNode(option, "computed class ineligible")
+			continue
+		case EvalComputedClassEscaped:
+			jobEscaped = true
+		}
+
+		// Run the job feasibility checks.
+		for _, check := range w.jobCheckers {
+			feasible := check.Feasible(option)
+			if !feasible {
+				// If the job hasn't escaped, set it to be ineligible since it
+				// failed a job check.
+				if !jobEscaped {
+					evalElig.SetJobEligibility(false, option.ComputedClass)
+				}
+				continue OUTER
+			}
+		}
+
+		// Set the job eligibility if the constraints weren't escaped.
+		if !jobEscaped {
```

Comment:
Probably want `if !jobEscaped && cached == EvalComputedClassUnknown` otherwise we are reseting job eligibility on every run.